organization/_posts/2018-03-21-packaging.md

```diff
@@ -0,0 +1,132 @@
+      - Marco - so then a binary cache is a specific install location cache
+      - /cvmfs/nix (or something like that) would be needed to share
+    - Patrick: Supercomputers use "uncvmfs"; user space filesystems also help (e.g., parrot)
+    - Total recompile does mean that it’s very clean, albeit with CPU cost for rebuilding
+    - You don’t need to worry about platform variants (SLC6, CentOS7, etc.) as single build is self contained.
+  - Package builds using Nix “Hydra” tool on openstack. Uses local machine for build, but easy to add additonal worker(s)
+    - Also supports for workers with different archs (darwin, or CPU variants etc)
+  - Overlays: Take main NixPkgs repo, then apply your changes on top of this. Can extend or override.
+    - Graeme: So can take things from base, but then change a few options? Yes.
+  - Environment Overlays
+    - "buildEnv" used to create environments. Symlinked to the store directory, similar to an LCG view.
+    - Also allows integrating with package overlays, and special treatment for languages that have their own conventions like Python/Haskell.
+- Examples given for installing Nix in Docker images, noted current limitations
+  - Binary downloads are slow
+  - Package signatures aren’t checked yet
+  - Some packages have issues when built inside Docker images (test suites that require privileges)
+- Why use nix?
+  - Built software should run on “any” Linux flavour (own libc, "build almost everything" approach)
+  - Simpler environments
+  - Many packages already upstream in NixPkgs
+  - Adding new packages is straight forward (many things like shbang patching, RUNPATH etc automagically handled)
+  - Active community
+- What isn’t so good
+  - Documentation lacking, but is improving and many examples help
+  - Nix language has a steep learning curve, but not so bad if just writing expressions
+  - Some issues with independence from host system
+    - Graphics, kernel (but no worse that others)
+  - Sometimes reproducible builds aren’t reproducible (mostly due to remote file URLs being changed)
+- Giulio
+  - Not easy to use “system” or “private” pieces, e.g., a local compiler
+  - Rebuild time is very long, if something changes
+    - Incremental build use case
+  - How to do my day to day development work?
+    - Especially changing some mid-level component, e.g. ROOT but checking impact on AliPhysics (without rebuilding everything?)
+      - AliBuild takes care of this
+    - Chris - Hydra will allow you to download a bootstrap script that gives a local build environment, that could be tweaked locally
+- Marco
+  - User development should be in nixshell
+  - At the moment we are trapped on a single well controlled environment (e.g., lxplus) as we can’t really support non-native systems
+  - Single tool to build and give a development environment is a big plus
+- Pere
+  - To build Gaudi source an lcgview, then cmake && make - there is nothing easier
```

Comment:
comma after Gaudi would make this clearer:

"To build Gaudi, source an lcgview..."